### PR TITLE
Enhancement: support dot-notated field properties in customapi widget

### PR DESCRIPTION
**Pull request by icyleaf**
## Proposed change

When [0.6.29](https://github.com/gethomepage/homepage/releases/tag/v0.6.29) with feature [CustomAPI](https://github.com/gethomepage/homepage/pull/1858) was released, It configure only in YAML file but invalid in docker tags.

Docker tags for example:

```yaml
labels:
    - homepage.group=Media
    - homepage.name=Emby
    - homepage.icon=emby.png
    - homepage.href=http://emby.home/
    - homepage.description=Media server
    - homepage.widget.type=customapi
    - homepage.widget.url=http://argus.service/api/v1/service/summary/emby
    - homepage.widget.field[0].label=Deployed Version
    - homepage.widget.field[0].field.status=deployed_version
    - homepage.widget.field[1].label=Latest Version
    - homepage.widget.field[1].field.status=latest_version
```

API response:

```json
{
  "id": "emby",
  "type": "github",
  "status": {
    "deployed_version": "4.8.0.35",
    "deployed_version_timestamp": "2023-10-16T06:39:44Z",
    "latest_version": "4.8.0.53",
    "latest_version_timestamp": "2023-10-16T06:39:44Z",
    "last_queried": "2023-10-16T06:39:44Z"
  }
}
```

## Type of change

- [ ] New service widget
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation only
- [ ] Other (please explain)

## Checklist:

- [ ] If adding a service widget or a change that requires it, I have added corresponding documentation changes.
- [ ] If adding a new widget I have reviewed the [guidelines](https://gethomepage.dev/latest/more/development/#service-widget-guidelines).
- [ ] If applicable, I have checked that all tests pass with e.g. `pnpm lint`.
- [ ] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.

### CODE REVIEW

File: docs/configs/docker.md [hunk 1]
```diff
@@ -153,6 +153,23 @@ labels:
     - homepage.widget.fields=["field1","field2"] # optional
 ```
 
+You can add fileds by using array-style dot notation with [CustomAPI](/widgets/services/customapi) widget.
```
Comment (shamoon) [inline docs/configs/docker.md:156]:
```suggestion
You can add specify fields for e.g. the [CustomAPI](/widgets/services/customapi) widget by using array-style dot notation:
```

**Review by shamoon — CHANGES_REQUESTED** [review]
So basically dot-prop handles arrays like this and shvl doesnt?

I will note https://bundlephobia.com/package/dot-prop@8.0.2 vs https://bundlephobia.com/package/shvl@3.0.0

**Comment by icyleaf** [comment]
> So basically dot-prop handles arrays like this and shvl doesnt?
> 
> I will note https://bundlephobia.com/package/dot-prop@8.0.2 vs https://bundlephobia.com/package/shvl@3.0.0

Yes, dot-prop accepts array-style parse, shvl not.

**Comment by benphelps** [comment]
> So basically dot-prop handles arrays like this and shvl doesnt?
> 
> I will note https://bundlephobia.com/package/dot-prop@8.0.2 vs https://bundlephobia.com/package/shvl@3.0.0

I checked the source for dot-prop and shvl, theres not much reason to have either as a dependency, here is a modification to shvl that fixes array style access:

```js
function get(object, path, def) {
  return (object = path.split(/[.[\]]+/).reduce(function (obj, p) {
    return obj && obj[p] !== undefined ? obj[p] : undefined;
  }, object)) === undefined ? def : object;
}

function set(object, path, val, obj) {
  path = path.split(/[.[\]]+/).filter((p) => p.length > 0);
  return ((path.slice(0, -1).reduce(function (obj, p) {
    return !/^(__proto__|constructor|prototype)$/.test(p) ? (obj[p] = obj[p] || {}) : {};
  }, (obj = object))[path.pop()] = val), object);
}
```

and testing it:

```js
let obj = { a: [{ a: "b" }, { a: "c" }] };
console.log(get(obj, "a[0].a")); // output "b"
set(obj, "a[1].a", "d"); 
console.log(get(obj, "a[1].a")); // output "d"
```

**Comment by icyleaf** [comment]
@benphelps function `set` failed by using shvl jest, and is there has a license issuse with modified code?

```diff
# "set(obj, 'x[0].y.one', 'foo')": { key: 'x[0].y.one', value: 'foo', expected: { x: [ { y: { one: 'foo' } } ] } }

- Expected  - 4
+ Received  + 4

  Object {
-   "x": Array [
-     Object {
+   "x": Object {
+     "0": Object {
        "y": Object {
-         "one": "bar",
+         "one": "foo",
        },
      },
-   ],
+   },
```

**Comment by benphelps** [comment]
Whoops, totally ignored the jest suite.  Fixed, passes all the tests plus new ones:

```js
export function get(object, path, def) {
  return (
    // Split the path into keys and reduce the object to the target value
    object = path.split(/[.[\]]+/).reduce(function (obj, p) {
      // Check each nested object to see if the key exists
      return obj && obj[p] !== undefined ? obj[p] : undefined;
    }, object)
  ) === undefined
    // If the final value is undefined, return the default value
    ? def
    : object;  // Otherwise, return the value found
}

export function set(obj, path, val) {
  // Split the path into keys and filter out any empty strings
  const keys = path.split(/[.[\]]+/).filter(Boolean);
  
  // Pop the last key to set the value later
  const lastKey = keys.pop();
  
  // Prevent setting dangerous keys like __proto__
  if (/^(__proto__|constructor|prototype)$/.test(lastKey)) return obj;

  // Reduce the object to the nested object where we want to set the value
  keys.reduce((acc, key, i) => {
    // Again, block dangerous keys
    if (/^(__proto__|constructor|prototype)$/.test(key)) return {};
    
    // Check if next key is an array index
    const isIndex = /^\d+$/.test(keys[i + 1]);
    
    // If current key doesn't exist, initialise it as an array or object
    acc[key] = Array.isArray(acc[key])
      ? acc[key]
      : (isIndex ? [] : acc[key] || {});
      
    // Return nested object for next iteration
    return acc[key];
  }, obj)[lastKey] = val;  // Finally set the value
  
  return obj;
}
```

It's not that much more code,  I commented it since it was so terse.

--- 

As for what to do with this... maybe `shvl` would accept a PR, but I doubt it.  We can include this along with the `shvl` MIT license in its own file and it'll be perfectly fine.

**Comment by shamoon** [comment]
I dont have super strong feelings but it is such a small use for this for us (like 4 spots) it does seem like keeping things as small as possible is better, plus Im sure ben had fun writing that patch =)

Agree license isn't an issue, just attribute the code and note MIT

**Comment by benphelps** [comment]
> I dont have super strong feelings but it is such a small use for this for us (like 4 spots) it does seem like keeping things as small as possible is better, plus Im sure ben had fun writing that patch =)
> 
> Agree license isn't an issue, just attribute the code and note MIT

Type II Fun... :D

Alright, @icyleaf, you're more than welcome to fold this patched shvl into the main code, again, just copy his MIT license over, with a link to the repo, and that should be good.  Or if you're not up to it, I'll grab this sometime tomorrow.

**Review by shamoon — APPROVED** [review]
LGTM

**Comment by cmh716** [comment]
I am struggling to get this working.  

API Response
{
    "name": "influxdb",
    "message": "ready for queries and writes",
    "status": "pass",
    "checks": [],
    "version": "v2.7.3",
    "commit": "ed645d9216"
}

Working in settings.yml:

          type: customapi
          url: http://192.168.0.60:8086/health
          method: GET
          mappings:
            - field: version
              label: Version
              format: text


Docker Labels:

        homepage.widget.type=customapi
        homepage.widget.url=http://192.168.0.60:8086/health
        homepage.widget.method=GET
        homepage.widget.field[0].field=version
        homepage.widget.field[0].label=Version


Any idea what I'm doing wrong?  Thanks for adding this feature.

**Comment by icyleaf** [comment]
@cmh716 I missed the `mappings` field, try this:

```
homepage.widget.mappings[0].field.field=version
homepage.widget.mappings[0].field.label=Version
```

**Comment by cmh716** [comment]
Thank you so much.  That didn't fix it, but you pointed me in the right direction.  This worked:

        homepage.widget.mappings[0].field=version
        homepage.widget.mappings[0].label=Version

**Pull request merged**